cogs/utils/p_strings.py

```diff
@@ -0,0 +1,285 @@
+
+
+def _get_pattern_from_string(string, anywhere=False):
+    """
+    Takes a string with placeholders and returns the regex pattern
+    By default, the pattern has start and end tokens, i.e. a string will have
+    to match the pattern exactly
+    If anywhere = True, the pattern doesn't have start/end tokens, i.e.
+    the pattern can be contained anywhere in the string
+    """
+    # for every %1%, %2%, etc... up to %9%, replace them by regex
+    # if-statement that will create named capture groups
+    # g1, g2, etc.. up to g9 on their first occurrence, and back-reference
+    # them on subsequent occurrences
+    for i in range(1, 10):
+        string = re.sub(f"%{i}%", f"(?(g{i})(?P=g{i})|(?P<g{i}>.+))", string)
+    # convert the choice lists to patterns
+    string = _convert_choice_list(string, to_pattern_str=True)
+    # if the string must be exactly the pattern, add start and end tokens
+    if not anywhere:
+        string = f"^{string}$"
+    # compile the regex pattern and return
+    return re.compile(string)
+
+
+class PString:
+    def __init__(self,
+                 string,
+                 user=None,
+                 channel=None,
+                 groups=None,
+                 additional_info=None):
+        """
+        A p-string is composed of a string with placeholders in it,
+        and values for these placeholders. Printing a p-string will
+        fill the string with these values.
+
+        Arguments:
+        - string: the string with placeholders
+
+        - user (placeholder: %user%): should generally be used with something
+        like ctx.author.mention but can be used with any string
+
+        - channel (placeholder: %channel%): should generally be used with
+        something like ctx.channel but can be used with any string
+
+        - groups (placeholder: %1%, %2%, etc... up to %9%): for example, if
+         there is a p-string with string "%1% likes %2%" and
+         groups=["He", "apples"], printing the p-string will return
+         "He likes apples"
+
+         - additional_info: to give more control to users, a p-string can store
+         additional information(this might be useful if a certain option is
+         selected when this p-string is called, etc.)
+
+        Other placeholder (not an argument):
+        - choice list (%[a,b]%): a random value will be chosen from the list
+        (here, either a or b). The values themselves can be choice lists
+        (ex: %[hi, hello %[world, you]%]%) will print either
+        "hi", "hello world" or "hello you"
+        """
+        self.string = string
+        self.user = user
+        self.channel = channel
+        self.groups = groups
+        self.additional_info = additional_info
+
+    def __str__(self):
+        filled_string = self.string
+        # replace the %user% and %channel% placeholders by their values
+        if self.user:
+            filled_string = re.sub("%user%", self.user, filled_string)
```

Comment:
i think we should precompile the `%user%` and `%channel%` regexes, as well as the %1% to %9% regexes

Reply:
Resolved in 8d3f102